app/src/mainwindow2.cpp

```diff
@@ -1511,10 +1566,115 @@ void MainWindow2::bindActionWithSetting(QAction* action, const SETTING& setting)
+                mToolLabel->setText(tr("Click to create a new polyline. Hold Ctrl and Shift to erase or Alt to select a color from the canvas."));
+            }
+            break;
+        case BUCKET:
+            mToolIcon->setPixmap(QPixmap(":icons/new/svg/bucket_detailed.svg"));
+            mToolLabel->setText(tr("Click to fill an area with the current color. Hold Alt to select a color from the canvas."));
+            break;
+        case EYEDROPPER:
+            mToolIcon->setPixmap(QPixmap(":icons/new/svg/eyedropper_detailed.svg"));
+            mToolLabel->setText(tr("Click to select a color from the canvas."));
+            break;
+        case BRUSH:
+            mToolIcon->setPixmap(QPixmap(":icons/new/svg/brush_detailed.svg"));
+            mToolLabel->setText(tr("Click to paint. Hold Ctrl and Shift to erase or Alt to select a color from the canvas."));
+            break;
+        default:
+            Q_ASSERT(false);
+    }
+    mToolIcon->setToolTip(BaseTool::TypeName(tool));
+}
+
+void MainWindow2::updateLayerStatus()
+{
+    QSignalBlocker b(mLayerBox);
+    mLayerBox->clear();
+    for (int i = 0; i < mEditor->layers()->count(); i++)
+    {
+        Layer *layer = mEditor->layers()->getLayer(i);
+        mLayerBox->addItem(layer->name());
+        switch (layer->type())
+        {
+            case Layer::BITMAP:
+                mLayerBox->setItemIcon(i, mBitmapIcon);
+                break;
+            case Layer::VECTOR:
+                mLayerBox->setItemIcon(i, mVectorIcon);
+                break;
+            case Layer::SOUND:
+                mLayerBox->setItemIcon(i, mSoundIcon);
+                break;
+            case Layer::CAMERA:
+                mLayerBox->setItemIcon(i, mCameraIcon);
+                break;
+            case Layer::MOVIE:
+            case Layer::UNDEFINED:
+                // no icon
+                break;
+        }
+    }
+    mLayerBox->setCurrentIndex(mEditor->layers()->currentLayerIndex());
+}
+
+void MainWindow2::updateLayerStatus(int layer)
+{
+    mLayerBox->setItemText(layer, mEditor->layers()->getLayer(layer)->name());
+    mLayerBox->setCurrentIndex(layer);
+}
+
+void MainWindow2::updateZoomStatus()
 {
-    qreal zoom = mEditor->view()->scaling() * 100.f;
-    statusBar()->showMessage(tr("Zoom: %0%").arg(zoom, 0, 'f', 1));
+    double zoom = mEditor->view()->scaling() * 100;
+    QSignalBlocker b1(mZoomBox);
+    mZoomBox->clear();
+    // Keep the dropdown list limited to our predefined entries
+    // insertPolicy can't handle that without preventing custom values outright
+    mZoomBox->addItems(QStringList() << "10000.0%" << "6400.0%" << "1600.0%" << "800.0%" << "400.0%" << "200.0%" << "100.0%" << "75.0%" << "50.0%" << "33.0%" << "25.0%" << "12.0%" << "1.0%");
+    mZoomBox->setCurrentText(QString("%0%").arg(zoom, 0, 'f', 1));
+    QSignalBlocker b2(mZoomSlider);
+    mZoomSlider->setValue(static_cast<int>(std::round(std::log10(mEditor->view()->scaling()) * 100)));
```

Comment:
`updateZoomStatus()` is supposed to be lightweight and fast. It will be triggered many times in a sec, for instance, when scrolling the mouse wheel. Try to avoid clearing and rebuilding dropdown list items here?

Reply:
Good point, thanks for catching that! Will think about another way.